### PR TITLE
Converted VariantCallingMetrics to using mergableMetricsBase 

### PR DESCRIPTION
**Pull request by yfarjoun**
- converting VariantCallingMetircs to mergable metrics
- added tests
- extended MergableMetricsBase so that it can deal with private and protected members of the superclass
- cleaned up alignment/typos in FilterVcf and MergeBamAlignment (unrelated)

@nh13 you has said that you'd like to see this framework used for something other than IndependentReplicates...so, here's your chance!  😄

### CODE REVIEW

**Comment by ktibbett** [comment]
@tfenne , @nh13  thinks that you are the one who asked to review this....

File: src/java/picard/analysis/replicates/MergeableMetricBase.java [hunk 1]
```diff
@@ -94,24 +98,34 @@ public boolean canMerge(final MergeableMetricBase other) {
...
     }
 
     /**
+     * for a collection of MergeableMetricBase, merge them all into "this" one.
+     *
+     * @param many a Collection of MergeableMetricBase
+     */
+    public MergeableMetricBase merge(final Collection<? extends MergeableMetricBase> many) {
+        many.parallelStream().forEach(this::merge);
+        calculateDerivedFields();
```
Comment (nh13) [inline src/java/picard/analysis/replicates/MergeableMetricBase.java]: but the `merge` below does not call `calculateDerivedFields`!  please remove.

File: src/java/picard/analysis/replicates/MergeableMetricBase.java [hunk 2]
```diff
@@ -121,6 +135,8 @@ public void merge(final MergeableMetricBase other) {
             }
 
             final Annotation[] summableAnnotations = field.getAnnotationsByType(MergeByAdding.class);
+            field.setAccessible(true);
```
Comment (nh13) [inline src/java/picard/analysis/replicates/MergeableMetricBase.java]: this would allow private and protected fields to merged?

File: src/java/picard/analysis/replicates/MergeableMetricBase.java [hunk 3]
```diff
@@ -179,11 +197,22 @@ public void merge(final MergeableMetricBase other) {
                 }
             }
         }
+        return this;
+    }
+
+    private static List<Field> getAllFields(Class clazz){
+        List<Field> fields = new ArrayList<>();
```
Comment (nh13) [inline src/java/picard/analysis/replicates/MergeableMetricBase.java]: final

File: src/java/picard/analysis/replicates/MergeableMetricBase.java [hunk 4]
```diff
@@ -179,11 +197,22 @@ public void merge(final MergeableMetricBase other) {
...
+        fields.addAll(Arrays.asList(clazz.getDeclaredFields()));
+        Class superClass = clazz.getSuperclass();
```
Comment (nh13) [inline src/java/picard/analysis/replicates/MergeableMetricBase.java]: final

File: src/java/picard/analysis/replicates/MergeableMetricBase.java [hunk 5]
```diff
@@ -179,11 +197,22 @@ public void merge(final MergeableMetricBase other) {
                 }
             }
         }
+        return this;
+    }
+
+    private static List<Field> getAllFields(Class clazz){
```
Comment (nh13) [inline src/java/picard/analysis/replicates/MergeableMetricBase.java]: final

File: src/java/picard/analysis/replicates/MergeableMetricBase.java [hunk 6]
```diff
@@ -179,11 +197,22 @@ public void merge(final MergeableMetricBase other) {
...
+
+        return fields;
     }
 
     /**
      * placeholder method that will calculate the derived fields from the other ones. classes that are derived from non-trivial base classes
      * should consider calling super.calculateDerivedFields() as well.
      */
-     public void calculateDerivedFields(){}
+     abstract public void calculateDerivedFields();
```
Comment (nh13) [inline src/java/picard/analysis/replicates/MergeableMetricBase.java]: Not all metrics will have derived fields, so I like the default implementation!

File: src/java/picard/sam/MergeBamAlignment.java [hunk 7]
```diff
@@ -146,11 +146,11 @@
 
     @Option(doc = "Reserved alignment attributes (tags starting with X, Y, or Z) that should be " +
             "brought over from the alignment data when merging.")
-    public List<String> ATTRIBUTES_TO_RETAIN = new ArrayList<String>();
+    public List<String> ATTRIBUTES_TO_RETAIN = new ArrayList<>();
 
     @Option(doc = "Attributes from the alignment record that should be removed when merging." +
             "  This overrides ATTRIBUTES_TO_RETAIN if they share common tags.")
-    public List<String> ATTRIBUTES_TO_REMOVE = new ArrayList<String>();
+    public List<String> ATTRIBUTES_TO_REMOVE = new ArrayList<>();
```
Comment (nh13) [inline src/java/picard/sam/MergeBamAlignment.java]: Good stuff, but could this at least be in it's own commit?

File: src/java/picard/vcf/filter/FilterVcf.java [hunk 8]
```diff
@@ -110,51 +110,49 @@ protected int doWork() {
...
+                if (sequenceDictionary == null) {
+                    throw new PicardException("The input vcf must have a sequence dictionary in order to create indexed vcf or bcfs.");
+                }
+                variantContextWriterBuilder.setReferenceDictionary(sequenceDictionary);
+            }
+            out = variantContextWriterBuilder.setOutputFile(OUTPUT).build();
+            header.addMetaDataLine(new VCFFilterHeaderLine("AllGtsFiltered", "Site filtered out because all genotypes are filtered out."));
+            header.addMetaDataLine(new VCFFormatHeaderLine("FT", VCFHeaderLineCount.UNBOUNDED, VCFHeaderLineType.String, "Genotype filters."));
+            for (final VariantFilter filter : variantFilters) {
+                filter.headerLines().forEach(header::addMetaDataLine);
```
Comment (nh13) [inline src/java/picard/vcf/filter/FilterVcf.java]: !

**Comment by nh13** [comment]
@yfarjoun rebase onto master and I can revisit.  The only comment that requires discussion is the one inside the merging of a collection (`calculateDerivedFields`).

File: src/java/picard/vcf/filter/FilterVcf.java (hunk 8 above)
Comment (yfarjoun) [inline src/java/picard/vcf/filter/FilterVcf.java]: what?

File: src/java/picard/sam/MergeBamAlignment.java (hunk 7 above)
Comment (nh13) [inline src/java/picard/sam/MergeBamAlignment.java]: Some comments are optional.  I tend to like separate commits for a clean history.  To each his/her own.

File: src/java/picard/vcf/filter/FilterVcf.java (hunk 8 above)
Comment (nh13) [inline src/java/picard/vcf/filter/FilterVcf.java]: Yeah not a good comment, since I don't remember.

**Comment by yfarjoun** [comment]
@nh13 do you have a few minutes to review this old PR?

File: src/main/java/picard/analysis/MergeableMetricBase.java [hunk 9]
```diff
@@ -213,6 +218,16 @@ public MergeableMetricBase merge(final MergeableMetricBase other) {
         return this;
     }
 
+    private static List<Field> getAllFields(Class clazz){
+        List<Field> fields = new ArrayList<>();
+        fields.addAll(Arrays.asList(clazz.getDeclaredFields()));
+        Class superClass = clazz.getSuperclass();
```
Comment (nh13) [inline src/main/java/picard/analysis/MergeableMetricBase.java]: finals?

File: src/main/java/picard/vcf/CollectVariantCallingMetrics.java [hunk 10]
```diff
@@ -231,72 +250,36 @@ public void updateDerivedValuesInPlace() {
...
          * Hidden fields not propagated to the metrics file.
          */
+        @MergeByAdding
         long numHets, numHomVar;
 
         public static String getFileExtension() {
             return "variant_calling_detail_metrics";
         }
 
-        public static void foldInto(final VariantCallingDetailMetrics target,
```
Comment (nh13) [inline src/main/java/picard/vcf/CollectVariantCallingMetrics.java]: Should be ok to delete instead of deprecate right?

**Comment by yfarjoun** [comment]
This fell through the cracks. Did the changes requested, will merge after tests pass.

**Comment by yfarjoun** [comment]
This fell through the cracks. Did the changes requested, will merge after tests pass.
